Origin: https://github.com/php/php-src/commit/7ac9700d0d46a7b96e72e330efd295a3332124b0
Advisory: GHSA-ch8v-r6jh-4vvr
Imported-by: deb.sury.org php7.4 7.4.33-31 0158, applies as-is to 7.3

From: Ilia Alshanetsky <ilia@ilia.ws>
Date: Wed, 29 Jul 2026 11:21:50 -0400
Subject: Fix GHSA-ch8v-r6jh-4vvr: encode 0xFF in FILTER_SANITIZE_ENCODED

php_filter_encode_url() initialized its 256-byte "must encode" table with
memset(tmp, 1, sizeof(tmp) - 1), leaving tmp[255] uninitialized. Whether
0xFF got percent-encoded then depended on stack garbage; valgrind reports
the read as a conditional jump on an uninitialised value. Initialize the
whole table.

Backport of 56149186d81469fd7fb2f8444359ee81dd4a70f4 from PHP-8.4.

(cherry picked from commit 7ac9700d0d46a7b96e72e330efd295a3332124b0)
(cherry picked from commit 9e4f091d31088af32da3c3247b1a8a51f703f80d)
(cherry picked from commit e1ae7c43933a8fb933beee712a3bc7f500ed7aae)
---
diff --git a/ext/filter/sanitizing_filters.c b/ext/filter/sanitizing_filters.c
index de69b3bf5a8b58ab05b11f6f81025f164cf36e92..b9d02d481758be37f858eb2c76ba5d1d6dcbe0f3 100644
--- a/ext/filter/sanitizing_filters.c
+++ b/ext/filter/sanitizing_filters.c
@@ -69,7 +69,7 @@ static void php_filter_encode_url(zval *value, const unsigned char* chars, const
 	unsigned char *e = s + char_len;
 	zend_string *str;
 
-	memset(tmp, 1, sizeof(tmp)-1);
+	memset(tmp, 1, sizeof(tmp));
 
 	while (s < e) {
 		tmp[*s++] = '\0';
diff --git a/ext/filter/tests/filter_sanitize_encoded_0xff.phpt b/ext/filter/tests/filter_sanitize_encoded_0xff.phpt
new file mode 100644
index 0000000000000000000000000000000000000000..1ee61aba7b70e921e518fcdb807772c1063d4200
--- /dev/null
+++ b/ext/filter/tests/filter_sanitize_encoded_0xff.phpt
@@ -0,0 +1,12 @@
+--TEST--
+FILTER_SANITIZE_ENCODED percent-encodes 0xFF
+--EXTENSIONS--
+filter
+--FILE--
+<?php
+var_dump(filter_var("\xFF", FILTER_SANITIZE_ENCODED));
+var_dump(filter_var("\xFE\xFF\x00A", FILTER_SANITIZE_ENCODED));
+?>
+--EXPECT--
+string(3) "%FF"
+string(10) "%FE%FF%00A"
